### PR TITLE
feat: Integrate `cspell` checker in CI for `/docs` and update PR template

### PR DESCRIPTION
**Pull request by dutterbutter**
**Series of PRs: This is the first in a series PRs aimed at enhancing spelling accuracy in this repository.** See:

- [ ] https://github.com/matter-labs/zksync-era/pull/682
- [ ] https://github.com/matter-labs/zksync-era/pull/683
- [ ] https://github.com/matter-labs/zksync-era/pull/684
- [ ] https://github.com/matter-labs/zksync-era/pull/685
- [ ] https://github.com/matter-labs/zksync-era/pull/827

Once merged, a final PR will enable the `dev_comments: true` in the cargo-spellcheck config file.

## What ❔

- **Introduction of cspell checker:** This PR integrates the cspell checker into the CI, specifically targeting the `/docs` directory to identify and correct spelling errors.
- Updates PR template to include `cspell` command 
- Updates dictionary

## Why ❔

- **Improving Documentation Quality:** The `/docs` directory currently contains numerous spelling errors. By resolving these and integrating a spellcheck step in our CI, we aim to maintain high-quality, error-free documentation.
- **Preventing Future Spelling Mistakes:** The cspell checker will help in automatically detecting and preventing spelling errors in future contributions.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `cargo spellcheck --cfg=./spellcheck/era.cfg --code 1`.

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -18,4 +18,5 @@
 - [ ] Tests for the changes have been added / updated.
 - [ ] Documentation comments have been added / updated.
 - [ ] Code has been formatted via `zk fmt` and `zk lint`.
-- [ ] Spellcheck has been run via `cargo spellcheck --cfg=./spellcheck/era.cfg --code 1`.
+- [ ] Spellcheck has been run via `cargo spellcheck --cfg=./spellcheck/era.cfg --code 1` & `cspell docs/\*_/_.md
```
Comment (Deniallugo) [inline .github/pull_request_template.md:21]: let's add it to `zk` tool

**Review by Deniallugo — COMMENTED** [review]
The only thing, i'm not sure that we have to fix spelling in rust code in docs.. 

From the one hand i'm pretty sure we should. 
From another hand if it has been taken from crypto repos, it could be confusing when reader will try to find it.. 
@mm-zk wdyt?

File: .github/pull_request_template.md (hunk 1 above)
Comment (dutterbutter) [reply .github/pull_request_template.md]: @Deniallugo added.

Comment (Deniallugo) [reply .github/pull_request_template.md]: Please update the template and github actions correspondingly

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -18,4 +18,4 @@
 - [ ] Tests for the changes have been added / updated.
 - [ ] Documentation comments have been added / updated.
 - [ ] Code has been formatted via `zk fmt` and `zk lint`.
-- [ ] Spellcheck has been run via `cargo spellcheck --cfg=./spellcheck/era.cfg --code 1`.
+- [ ] Spellcheck has been run via `cargo spellcheck --cfg=./spellcheck/era.cfg --code 1` & `zk cspell`
```
Comment (Deniallugo) [inline .github/pull_request_template.md:21]: why `&` ? I'd assume it's supposed  to be the part of zk spell

File: infrastructure/zk/src/spellcheck.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        }
+    } catch (error) {
+        console.error('Error occurred during spell checking:', error);
+    }
+}
+
+export const command = new Command('spellcheck')
+    .option('--pattern <pattern>', 'Glob pattern for files to check', 'docs/**/*')
+    .option('--config <config>', 'Path to configuration file', './spellcheck/cspell.json')
+    .option('--use-cargo', 'Use cargo spellcheck')
```
Comment (Deniallugo) [inline infrastructure/zk/src/spellcheck.ts:32]: can we use both by default?

Comment (dutterbutter) [reply infrastructure/zk/src/spellcheck.ts:32]: It does run both by default. Running `zk spellcheck` will run both cspell for docs, and cargo spellcheck. You also have the option to do them separately running `zk spellcheck --use-cspell` and `zk spellcheck --use-cargo`.

**Review by Deniallugo — APPROVED** [review]

File: .github/workflows/check-spelling.yml [hunk 4]
```diff
@@ -23,3 +23,32 @@ jobs:
 
       - name: Run cargo-spellcheck
         run: cargo spellcheck --cfg=./spellcheck/era.cfg --code 1
```
Comment (Deniallugo) [inline .github/workflows/check-spelling.yml:25]: Are you sure we need both? 
could we just run `zk spellcheck`?

Comment (dutterbutter) [reply .github/workflows/check-spelling.yml:25]: Only reason I separated them was the I kept getting a failure when installing `cargo-spellcheck` on the matterlabs-ci-runner. So I separated so spell-check was installed and ran on ubuntu-latest.

**Comment by dutterbutter** [comment]
@matter-labs/devops could you assist me with the CI here? Inspecting the `check-spelling` workflow logs it looks like `/bin/sh: 1: cspell: not found` . The package `cspell` has been added in the `package.json` for the zk tool to use in this PR but doesnt seem to be getting picked up during the workflow run?

File: infrastructure/zk/package.json [hunk 5]
```diff
@@ -32,6 +32,6 @@
         "@types/tabtab": "^3.0.1",
         "hardhat": "=2.16.0",
         "typescript": "^4.3.5",
-        "cspell": "^8.2.3"
+        "cspell": "*"
```
Comment (Deniallugo) [inline infrastructure/zk/package.json:35]: i'd prefer to specify the version

File: infrastructure/zk/src/spellcheck.ts [hunk 6]
```diff
@@ -1,28 +1,34 @@
...
+
+        // Check results and exit with error code if any command failed
+        if (results.some((code) => code !== 0)) {
+            console.error('Spell check failed');
+            process.exit(1);
+        }
     } catch (error) {
         console.error('Error occurred during spell checking:', error);
+        process.exit(1);
     }
```
Comment (Deniallugo) [inline infrastructure/zk/src/spellcheck.ts:32]: do you even need this try catch? 
i don't see much value here

File: infrastructure/zk/package.json (hunk 5 above)
Comment (dutterbutter) [reply infrastructure/zk/package.json:35]: The reason I dont is that we are using at `latest` in the dockerfile so to keep consistency between CI environment and what it is reporting and local development.

Comment (Deniallugo) [reply infrastructure/zk/package.json:35]: ok

File: infrastructure/zk/src/spellcheck.ts (hunk 6 above)
Comment (dutterbutter) [reply infrastructure/zk/src/spellcheck.ts:32]: Yes its needed to correctly handle rejected promises and properly displays the error.

Comment (Deniallugo) [reply infrastructure/zk/src/spellcheck.ts:32]: `Error occurred during spell checking: Child process exited with code 1` 

Are you sure?
If it's really necessary I don't mind

**Review by Deniallugo — APPROVED** [review]

**Review by alexandrst88 — APPROVED** [review]

**Pull request merged**